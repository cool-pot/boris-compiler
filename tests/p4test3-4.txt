*** test level 3,  local foreach array
defun arrayfunc(x)
    local myarray;
    array myarray[5..10];
    local i = 5;
    while i <= 10 do 
        myarray[i] = 2*i;
        i = i+1;
    end while
    foreach k in myarray do
        print k;
    end for
    return 0;
end defun

print arrayfunc 0;