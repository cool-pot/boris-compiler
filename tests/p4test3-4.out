; ModuleID = 'module-boris-input'
source_filename = "module-boris-input"
target triple = "x86_64-pc-linux-gnu"

@format_int = private unnamed_addr constant [4 x i8] c"%d\0A\00"
@format_int.1 = private unnamed_addr constant [4 x i8] c"%d\0A\00"

define i32 @strlen(i8* %s) {
init:
  %i = alloca i32
  store i32 0, i32* %i
  br label %check

check:                                            ; preds = %body, %init
  %0 = load i32, i32* %i
  %1 = getelementptr i8, i8* %s, i32 %0
  %2 = load i8, i8* %1
  %3 = icmp ne i8 0, %2
  br i1 %3, label %body, label %end

body:                                             ; preds = %check
  %4 = load i32, i32* %i
  %5 = add i32 %4, 1
  store i32 %5, i32* %i
  br label %check

end:                                              ; preds = %check
  %6 = load i32, i32* %i
  ret i32 %6
}

declare i32 @printf(...)

define void @main() {
MainEntry:
  %0 = call i32 @arrayfunc(i32 0)
  %1 = call i32 (...) @printf(i8* getelementptr inbounds ([4 x i8], [4 x i8]* @format_int.1, i32 0, i32 0), i32 %0)
  br label %EndEntry

EndEntry:                                         ; preds = %MainEntry
  ret void
}

define i32 @arrayfunc(i32) {
Entry:
  %x = alloca i32
  store i32 %0, i32* %x
  %myarray = alloca i32, i32 6
  %i = alloca i32
  store i32 5, i32* %i
  br label %while_init_

while_init_:                                      ; preds = %Entry
  br label %while_check_

while_check_:                                     ; preds = %while_body_, %while_init_
  %load_singleid_int_temp = load i32, i32* %i
  %1 = icmp sle i32 %load_singleid_int_temp, 10
  br i1 %1, label %while_body_, label %while_end_

while_body_:                                      ; preds = %while_check_
  %load_singleid_int_temp1 = load i32, i32* %i
  %2 = sub i32 %load_singleid_int_temp1, 5
  %3 = getelementptr i32, i32* %myarray, i32 %2
  %load_singleid_int_temp2 = load i32, i32* %i
  %expr_mul_expr_tmp = mul i32 2, %load_singleid_int_temp2
  store i32 %expr_mul_expr_tmp, i32* %3
  %load_singleid_int_temp3 = load i32, i32* %i
  %expr_plus_expr_tmp = add i32 %load_singleid_int_temp3, 1
  store i32 %expr_plus_expr_tmp, i32* %i
  br label %while_check_

while_end_:                                       ; preds = %while_check_
  br label %foreacha_init_

foreacha_init_:                                   ; preds = %while_end_
  %_temp_id_0 = alloca i32
  %k = alloca i32
  store i32 0, i32* %_temp_id_0
  %load_singleid_int_temp4 = load i32, i32* %_temp_id_0
  %4 = getelementptr i32, i32* %myarray, i32 %load_singleid_int_temp4
  %5 = load i32, i32* %4
  store i32 %5, i32* %k
  br label %foreacha_check_

foreacha_check_:                                  ; preds = %foreacha_body_, %foreacha_init_
  %load_singleid_int_temp5 = load i32, i32* %_temp_id_0
  %6 = icmp slt i32 %load_singleid_int_temp5, 6
  br i1 %6, label %foreacha_body_, label %foreacha_end_

foreacha_body_:                                   ; preds = %foreacha_check_
  %load_singleid_int_temp6 = load i32, i32* %k
  %7 = call i32 (...) @printf(i8* getelementptr inbounds ([4 x i8], [4 x i8]* @format_int, i32 0, i32 0), i32 %load_singleid_int_temp6)
  %load_singleid_int_temp7 = load i32, i32* %_temp_id_0
  %8 = add i32 %load_singleid_int_temp7, 1
  store i32 %8, i32* %_temp_id_0
  %load_singleid_int_temp8 = load i32, i32* %_temp_id_0
  %9 = getelementptr i32, i32* %myarray, i32 %load_singleid_int_temp8
  %10 = load i32, i32* %9
  store i32 %10, i32* %k
  br label %foreacha_check_

foreacha_end_:                                    ; preds = %foreacha_check_
  ret i32 0
}
